### PR TITLE
C++ Rewrite

### PR DESCRIPTION
**Pull request by SzilBalazs**

### CODE REVIEW

File: knpy/braid_vec.py [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+        sigmas: Braid notation, e.g. [1,-1,2] or the string name of knot e.g. 4_1 #TODO Above 10 there a and n knots
+            (e.g. 11a,13n)
+        notation_index: If sigmas is a name of braid than it is possible that multiple notations are available to the
+            same knot. notation_index says which one to choose from these.
+        #TODO 10_136 {{-1;-1;-2;3;-2;1;-2;-2;3;2;2};{-1;2;-1;2;3;-2;-2;-4;3;-4}}? Which one?
+        #TODO 11n_8,{{-1;-1;-2;1;-2;-1;3;-2;-2;-4;3;-4};{1;2;-1;2;3;-2;-1;-1;-2;-2;-3;-3;-2}}? Which one?
+        """
+        self._braid: BraidNotation
+        if isinstance(sigmas, str):
+            self._braid = np.array(knots_in_braid_notation_dict[sigmas][notation_index])
```
Comment (Error-42) [inline knpy/braid_vec.py:46]: `dtype` should be set to `np.int32`, same as in `braid.py`.

File: knpy/braid_vec.py [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+            if copy_sigmas:
+                self._braid = sigmas.copy()
+            else:
+                self._braid = sigmas
+        else:
+            if not all(isinstance(x, (int, np.integer)) for x in sigmas):
+                raise InvalidBraidException(
+                    f"Unable to create braid from {type(sigmas)}, an element is not instance of int or np.integer"
+                )
+            self._braid = np.array(sigmas)
```
Comment (Error-42) [inline knpy/braid_vec.py:57]: `dtype` should be set to `np.int32`, same as in `braid.py`.

File: tests/test_braid_vec.py [hunk 3]
```diff
@@ -0,0 +1,684 @@
...
+from knpy.braid_vec import Braid
+# from knpy.braid import BraidTransformation
+from knpy import IllegalTransformationException, InvalidBraidException, IndexOutOfRangeException
+
+
+class TestBraidClassBraidRelationsInit:
+    def test_init_empty(self) -> None:
+        braid = Braid([])
+        assert braid.strand_count == 1
+        assert braid.notation().shape[0] == 0
```
Comment (Error-42) [inline tests/test_braid_vec.py:15]: Please apply changes from merges also applied to `test_braid.py`. Here: update tests to include types.

File: tests/test_braid_vec.py [hunk 4]
```diff
@@ -0,0 +1,684 @@
+from functools import partial
+import pytest
+import numpy as np
+
+# IMPORTANT: knpy should be installed first
+from knpy.braid_vec import Braid
+# from knpy.braid import BraidTransformation
```
Comment (Error-42) [inline tests/test_braid_vec.py:7]: `BraidTransformation` is later used where it is not defined.

File: knpy/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,15 @@
+cmake_minimum_required(VERSION 3.15)
```
Comment (Error-42) [inline knpy/CMakeLists.txt:1]: To not delay this review further, the general build process has not been reviewed. Although based on my experience of trying to do it on a Windows machine (and that you had to modify this file), it is not ready for review. If really needed, PR could be merged without proper building documentation and process.

File: README.md [hunk 6]
```diff
@@ -26,4 +26,39 @@ Push the branch:
...
\ No newline at end of file
+Create a pull request and wait for review.
+
+## Faster `Braid` implementation
+
+If you want to use the faster `Braid` implementation in `braid_vec.py`, you need 
+to compile the C++ modules.
+
+> ⚠️ Warning: although the APIs are very similar, some features, such as negative 
+indexing is not supported in the faster implementations.
```
Comment (Error-42) [inline README.md:37]: All lines of blockquotes should have `>` prefixed.

**Review by Error-42 — CHANGES_REQUESTED** [review]
Behaviour has been changed; negative indexes are less supported, exceptions catch fewer errors. Better documentation should be provided about this, make issues if merged. Also make issues, about behaviour differences, eventually tests should be merged.

**Review by Error-42 — CHANGES_REQUESTED** [review]
Please fix mypy errors.

**Comment by SzilBalazs** [comment]
> Please fix mypy errors.

```
knpy/braid_vec.py:10: error: Module "knpy" has no attribute "braid_cpp_impl"  [attr-defined]
knpy/__init__.py:7: error: Incompatible import of "Braid" (imported name has type "type[knpy.braid.Braid]", local name has type "type[knpy.braid_vec.Braid]")  [assignment]
```

I don't see any easy fixes, @sulan what do you think?

File: knpy/__init__.py [hunk 7]
```diff
@@ -1,2 +1,7 @@
-from .braid import Braid
+import os as _os
+
+if _os.environ.get("KNPY_FAST_BRAID", "no").lower() in ["on", "yes", "true", "1"]:
```
[the remark this answers is not in the document]
Comment (SzilBalazs) [reply knpy/__init__.py]: Done

File: knpy/braid.cpp [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+
+namespace py = pybind11;
+using array = py::array_t<long long>;
+
+struct IllegalTransformationException : public std::runtime_error {
+    using std::runtime_error::runtime_error;
+};
+
+int sign(int x) {
+    return x > 0 ? 1 : -1;
```
[the remark this answers is not in the document]
Comment (SzilBalazs) [reply knpy/braid.cpp:15]: Done

File: knpy/braid.cpp [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+}
+
+array shift_left(array _inp, int amount = 1) {
```
[the remark this answers is not in the document]
Comment (SzilBalazs) [reply knpy/braid.cpp]: Done, hopefully I didn't miss anything.

File: knpy/braid.cpp [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+    auto inp = _inp.unchecked<1>();
+    int n = inp.size();
+    array _res(n);
+    auto res = _res.mutable_unchecked();
```
[the remark this answers is not in the document]
Comment (SzilBalazs) [reply knpy/braid.cpp]: Done

File: knpy/braid_vec.py [hunk 11]
```diff
@@ -0,0 +1,370 @@
...
+        return Braid._from_array_directly(transformed)
+
+    # Chech whether a move is performable or not
+    def is_braid_relation1_performable(self, index: int) -> bool:
+        """
+        Check if braid relation 1 is performable at the index. See documentation of `braid_relation1` for details.
+
+        index: Where the chunk would start; in the range [-n, n) where n is the number of crossings in the braid (so n =
+            len(braid))
+        """
```
[the remark this answers is not in the document]
Comment (SzilBalazs) [reply knpy/braid_vec.py:244]: Done

File: knpy/braid.cpp [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+        return _res;
+    } else {
+        throw IllegalTransformationException("Illegal transformation");
+    }
+}
+
+bool is_braid_relation2_performable(array _inp, int index) {
+    auto inp = _inp.unchecked<1>();
+    int n = inp.size();
+    int i = (index+1)%n;
```
[the remark this answers is not in the document]
Comment (SzilBalazs) [reply knpy/braid.cpp]: Done

File: knpy/braid_vec.py [hunk 13]
```diff
@@ -0,0 +1,370 @@
...
+
+        Maps between the chunks `[i, j] ↔ [j, i]` if `abs(abs(i) - abs(j)) >= 2`.
+
+        The elements of the chuck must be distict, so the braid must consist of at least 2 crossings. The braid is
+        assumed to be circular, the chunk may cross the end of the array (so some elements from the end, then some
+        elements from the beginning).
+
+        index: Where the chunk starts, on which operation can be done; must be in the range [-n, n) where n is the
+            number of crossings in the braid (so n = len(braid))
+        """
```
[the remark this answers is not in the document]
Comment (SzilBalazs) [reply knpy/braid_vec.py:175]: Done

File: knpy/braid.cpp [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+        }
+    }
+    return _res;
+}
+
+bool is_destabilization_performable(array _inp, int index, int strand_count) {
+    auto inp = _inp.unchecked<1>();
+    int n = inp.size();
+    bool valid_index = 0 <= index && index < n;
+    bool ok1 = false, ok2 = true, ok3 = false, ok4 = true;
```
[the remark this answers is not in the document]
Comment (SzilBalazs) [reply knpy/braid.cpp]: Renamed variables

File: knpy/braid.cpp [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+    array _res(n);
+    auto res = _res.mutable_unchecked();
+    for (int i = 0; i < n; i++) res[i] = is_braid_relation2_performable(_inp, i);
+    return _res;
+}
+
+array braid_relation2(array _inp, int index) {
+    auto inp = _inp.unchecked<1>();
+    int n = inp.size();
+    if (is_braid_relation2_performable(_inp, index)) {
```
[the remark this answers is not in the document]
Comment (SzilBalazs) [reply knpy/braid.cpp]: Done

File: knpy/braid.cpp [hunk 16]
```diff
@@ -0,0 +1,249 @@
...
+            else ok4 = false;
+        }
+    }
+    return valid_index && ((ok1 && ok2) || (ok3 && ok4));
+}
+
+array destabilization(array _inp, int index, int strand_count) {
+    auto inp = _inp.unchecked<1>();
+    int n = inp.size();
+    if (is_destabilization_performable(_inp, index, strand_count)) {
```
[the remark this answers is not in the document]
Comment (SzilBalazs) [reply knpy/braid.cpp]: Done

File: knpy/braid.cpp [hunk 17]
```diff
@@ -0,0 +1,249 @@
...
+        bool on_top = abs(inp[index]) == 1;
+        array _res(n-1);
+        auto res = _res.mutable_unchecked();
+        for (int i = 0; i<n-1; i++) {
+            int j = i+(i>=index);
+            res[i] = inp[j] - on_top * (sign(inp[j]));
```
[the remark this answers is not in the document]
Comment (SzilBalazs) [reply knpy/braid.cpp]: Done

File: knpy/braid_vec.py [hunk 18]
```diff
@@ -0,0 +1,371 @@
...
+        return True
+
+    def is_destabilization_performable(self, index: int) -> bool:
+        """
+        Helper function to determine if destabilisation move is performable
+        at given index location, at either the top or bottom strand.
+        """
+        return B.is_destabilization_performable(self._braid, index, self.strand_count)
+
+    def is_remove_sigma_inverse_pair_performable(self, index: int) -> bool:
```
[the remark this answers is not in the document]
Comment (SzilBalazs) [reply knpy/braid_vec.py:300]: Done

File: knpy/braid_vec.py [hunk 19]
```diff
@@ -0,0 +1,371 @@
...
+        return Braid._from_array_directly(transformed)
+
+    @braid_move
+    def remove_sigma_inverse_pair(self, index: int) -> "Braid":
+        """
+        Remove consequtive inverse on a given place (last and first element are consequtive)
+
+        index: the smaller index (or when the pair is the last and first then it is the last index); must be in the
+            range [-k, k) where k is the number of crossings (so `len(braid)`).
+        """
```
[the remark this answers is not in the document]
Comment (SzilBalazs) [reply knpy/braid_vec.py:233]: Done

File: knpy/braid.cpp [hunk 20]
```diff
@@ -0,0 +1,249 @@
...
+    }
+}
+
+
+PYBIND11_MODULE(braid_cpp_impl, m) {
+    m.doc() = "Braid C++ implementation";
+    py::register_exception<IllegalTransformationException>(
+        m, "IllegalTransformationException");
+
+    m.def("shift_left", &shift_left, "Shift left implementation");
```
[the remark this answers is not in the document]
Comment (SzilBalazs) [reply knpy/braid.cpp:236]: Added a comment

File: knpy/braid_vec.py (hunk 1 above)
Comment (SzilBalazs) [reply knpy/braid_vec.py]: Done

File: tests/test_braid_vec.py (hunk 3 above)
Comment (SzilBalazs) [reply tests/test_braid_vec.py]: Done

File: tests/test_braid_vec.py (hunk 4 above)
Comment (SzilBalazs) [reply tests/test_braid_vec.py]: Done

File: README.md (hunk 6 above)
Comment (SzilBalazs) [reply README.md]: Done

File: knpy/braid.cpp [hunk 21]
```diff
@@ -0,0 +1,249 @@
...
+    }
+    return _res;
+}
+
+bool is_destabilization_performable(array _inp, int index, int strand_count) {
+    auto inp = _inp.unchecked<1>();
+    int n = inp.size();
+    bool valid_index = 0 <= index && index < n;
+    bool ok1 = false, ok2 = true, ok3 = false, ok4 = true;
+    int cnt = strand_count - 1;
```
[the remark this answers is not in the document]
Comment (SzilBalazs) [reply knpy/braid.cpp]: Done

File: knpy/braid.cpp [hunk 22]
```diff
@@ -0,0 +1,249 @@
...
+        for (int i = 0; i < 3; i++) {
+            if (inp[(index+i)%n] < 0) signs[i] = -1;
+        }
+        for (int i = 0; i < 3; i++) {
+            int j = (index+(i!=1))%n;
+            res[(index+i)%n] = signs[2-i] * abs(inp[j]); 
+        }
+        return _res;
+    } else {
+        throw IllegalTransformationException("Illegal transformation");
```
[the remark this answers is not in the document]
Comment (SzilBalazs) [reply knpy/braid.cpp]: Done

File: knpy/braid.cpp [hunk 23]
```diff
@@ -0,0 +1,249 @@
...
+    auto inp = _inp.unchecked<1>();
+    int n = inp.size();
+    if (is_braid_relation2_performable(_inp, index)) {
+        array _res(n);
+        auto res = _res.mutable_unchecked();
+        for (int i = 0; i < n; i++) res[i] = inp[i];
+        std::swap(res[index], res[(index+1)%n]);
+        return _res;
+    } else {
+        throw IllegalTransformationException("Illegal transformation");
```
[the remark this answers is not in the document]
Comment (SzilBalazs) [reply knpy/braid.cpp]: Done

File: knpy/braid.cpp [hunk 24]
```diff
@@ -0,0 +1,249 @@
...
+        bool on_top = abs(inp[index]) == 1;
+        array _res(n-1);
+        auto res = _res.mutable_unchecked();
+        for (int i = 0; i<n-1; i++) {
+            int j = i+(i>=index);
+            res[i] = inp[j] - on_top * (sign(inp[j]));
+        }
+        return _res;
+    } else {
+        throw IllegalTransformationException("Illegal transformation");
```
[the remark this answers is not in the document]
Comment (SzilBalazs) [reply knpy/braid.cpp]: Done

File: knpy/CMakeLists.txt (hunk 5 above)
Comment (SzilBalazs) [reply knpy/CMakeLists.txt:1]: I think let us proceed with the merge and revisit this problem at a later date.

File: knpy/braid.cpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (Error-42) [reply knpy/braid.cpp:238]: What do you mean by implicit parameter?

Also this comment doesn't solve the problem that the implicit constructor calls are hard to follow. (If I'm correct there are implicit constructor calls.)

File: knpy/braid.cpp [hunk 25]
```diff
@@ -0,0 +1,253 @@
...
+#include <utility>
+
+namespace py = pybind11;
+using array = py::array_t<long long>;
+
+struct IllegalTransformationException : public std::runtime_error {
+    using std::runtime_error::runtime_error;
+};
+
+int sign_of_non_zero(int x) {
```
Comment (Error-42) [inline knpy/braid.cpp:15]: Mark const.

Comment (SzilBalazs) [reply knpy/braid.cpp]: Marked

File: knpy/braid_vec.py [hunk 26]
```diff
@@ -0,0 +1,366 @@
...
+
+    # Action functions from paper https://arxiv.org/pdf/2010.16263
+
+    @braid_move
+    def shift_left(self, amount: int = 1) -> "Braid":
+        """
+        Shifts the crossings of the braid left. Numbering the original crossings as `[0, 1, 2, ..., n - 1]` it
+        transforms it to `[amount, amount + 1, amount + 2, ..., n - 2, n - 1, 0, 1, 2, ..., amount - 1]`.
+
+        amount: in the range (-n, n) where n is the number of crossings in the braid (so n = len(braid))
```
Comment (Error-42) [inline knpy/braid_vec.py:116]: Forgot to update docs, `amount` can't be negative.

File: knpy/braid_vec.py [hunk 27]
```diff
@@ -0,0 +1,366 @@
...
+                raise IllegalTransformationException("Cannot shift empty braid.")
+            raise IndexOutOfRangeException(f"Amount ({amount}) should be less than the length.")
+        shifted = B.shift_left(self._braid, amount)
+        return Braid._from_array_directly(shifted)
+
+    @braid_move
+    def shift_right(self, amount: int = 1) -> "Braid":
+        """
+        Shifts the crossings of the braid right. Same as shifting left by the negative amount.
+        amount: in the range (-n, n) where n is the number of crossings in the braid (so n = len(braid))
```
Comment (Error-42) [inline knpy/braid_vec.py:131]: Forgot to update docs, `amount` can't be negative.

File: knpy/braid_vec.py [hunk 28]
```diff
@@ -0,0 +1,366 @@
...
+        return np.nonzero(B.braid_relation1_performable_indices(self._braid))[0]
+
+    def is_braid_relation2_performable(self, index: int) -> bool:
+        if index >= len(self._braid):
+            raise IndexOutOfRangeException(
+                f"index = {index} too large, at least the number of crossings = {len(self._braid)}"
+            )
+        if index < 0:
+            raise IndexOutOfRangeException(
+                f"index = {index} too small, smaller than 0 = {-len(self._braid)}"
```
Comment (Error-42) [inline knpy/braid_vec.py:263]: `-len(self.braid)` is not necessarily 0.

File: knpy/braid_vec.py [hunk 29]
```diff
@@ -0,0 +1,366 @@
...
+
+    # Action functions from paper https://arxiv.org/pdf/2010.16263
+
+    @braid_move
+    def shift_left(self, amount: int = 1) -> "Braid":
+        """
+        Shifts the crossings of the braid left. Numbering the original crossings as `[0, 1, 2, ..., n - 1]` it
+        transforms it to `[amount, amount + 1, amount + 2, ..., n - 2, n - 1, 0, 1, 2, ..., amount - 1]`.
+
+        amount: in the range (0, n) where n is the number of crossings in the braid (so n = len(braid))
```
Comment (Error-42) [inline knpy/braid_vec.py:116]: Amount can be zero.

**Review by Error-42 — APPROVED** [review]
Known issues:

- Trying to setup development is hard, manual help is required.
- Tests had to be modified.
- Tests are duplicated for implementations.
- Mypy throws an error, because mypy can't work out stuff
